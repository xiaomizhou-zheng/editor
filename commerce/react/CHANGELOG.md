# @af/commerce-react

## 0.0.4

### Patch Changes

- [`8cc2f888c83`](https://bitbucket.org/atlassian/atlassian-frontend/commits/8cc2f888c83) - Upgrade Typescript from `4.3.5` to `4.5.5`

## 0.0.3

### Patch Changes

- [`8d4228767b0`](https://bitbucket.org/atlassian/atlassian-frontend/commits/8d4228767b0) - Upgrade Typescript from `4.2.4` to `4.3.5`.

## 0.0.2

### Patch Changes

- [`cb2392f6d33`](https://bitbucket.org/atlassian/atlassian-frontend/commits/cb2392f6d33) - Upgrade to TypeScript 4.2.4

## 0.0.1

### Patch Changes

- [`e5c858ab40c`](https://bitbucket.org/atlassian/atlassian-frontend/commits/e5c858ab40c) - Updating stripe and managing state update

## 0.0.4

### Patch Changes

- [`229b32842b5`](https://bitbucket.org/atlassian/atlassian-frontend/commits/229b32842b5) - Fix .npmignore and tsconfig.json for **tests**

## 0.0.3

### Patch Changes

- [`d3265f19be`](https://bitbucket.org/atlassian/atlassian-frontend/commits/d3265f19be) - Transpile packages using babel rather than tsc

## 0.0.2

### Patch Changes

- [`5f58283e1f`](https://bitbucket.org/atlassian/atlassian-frontend/commits/5f58283e1f) - Export types using Typescript's new "export type" syntax to satisfy Typescript's --isolatedModules compiler option.
  This requires version 3.8 of Typescript, read more about how we handle Typescript versions here: https://atlaskit.atlassian.com/get-started
  Also add `typescript` to `devDependencies` to denote version that the package was built with.

## 0.0.1

### Patch Changes

- [`b443b5a60f`](https://bitbucket.org/atlassian/atlassian-frontend/commits/b443b5a60f) - Renamed template package
