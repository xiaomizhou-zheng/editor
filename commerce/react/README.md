# CommerceReact

undefined

## Usage

`import CommerceReact from '@atlaskit/commerce-react';`

Detailed docs and example usage can be found [here](https://atlaskit.atlassian.com/packages/commerce/commerce-react).
