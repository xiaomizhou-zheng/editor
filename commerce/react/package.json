{
  "name": "@atlaskit/commerce-react",
  "version": "0.0.4",
  "author": "Atlassian Pty Ltd",
  "license": "Apache-2.0",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/"
  },
  "atlassian": {
    "team": "Commerce",
    "inPublicMirror": false,
    "releaseModel": "continuous",
    "website": {
      "name": "CommerceReact"
    }
  },
  "repository": "https://bitbucket.org/atlassian/atlassian-frontend",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "module:es2019": "dist/es2019/index.js",
  "types": "dist/types/index.d.ts",
  "sideEffects": false,
  "atlaskit:src": "src/index.ts",
  "af:exports": {
    ".": "./src/index.ts"
  },
  "dependencies": {
    "@babel/runtime": "^7.0.0"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  },
  "devDependencies": {
    "@atlaskit/docs": "*",
    "@atlassian/atlassian-frontend-prettier-config-1.0.1": "npm:@atlassian/atlassian-frontend-prettier-config@1.0.1",
    "@testing-library/react": "^12.1.5",
    "react-dom": "^16.8.0",
    "typescript": "4.5.5",
    "wait-for-expect": "^1.2.0"
  },
  "techstack": {
    "@atlassian/frontend": {
      "code-structure": [
        "tangerine-next"
      ],
      "import-structure": [
        "atlassian-conventions"
      ],
      "circular-dependencies": [
        "file-and-folder-level"
      ]
    },
    "@repo/internal": {
      "analytics": [
        "analytics-next"
      ],
      "theming": [
        "react-context"
      ],
      "ui-components": [
        "lite-mode"
      ],
      "deprecation": [
        "no-deprecated-imports"
      ]
    }
  },
  "prettier": "@atlassian/atlassian-frontend-prettier-config-1.0.1"
}
