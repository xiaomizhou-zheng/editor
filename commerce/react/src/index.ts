export { batchedUpdate } from './common/utils/batch';
