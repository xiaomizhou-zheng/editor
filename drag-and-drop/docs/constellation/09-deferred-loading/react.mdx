---
order: 1
title: Deferred loading
description: Improving performing by delaying the loading of drag and drop
---

You can use _dynamic imports_ to do things like deferring the import of a module until a `react` `useEffect`.

## Basic usage

```tsx
import React, {useEffect} from 'react';
import invariant from 'tiny-invariant';

function Card() {
  const ref = useRef<HTMLDivElement | null>(null);
  const [state, setState] = useState();

  useEffect(() => {
    const controller = new AbortController();

    (async () => {
      // load in all the modules that you need
      const modules = await Promise.all([
        import('@atlaskit/drag-and-drop/adapter/element'),
        import('@atlaskit/drag-and-drop/util/combine'),
      ]);
      if (controller.signal.aborted) {
        return;
      }
      invariant(el);

      const [{draggable, dropTargetForElements}, {combine}] = modules;

      const cleanup = combine(
        draggable({
        element: el,
        onDragStart: () => setState('dragging'),
        onDrop: () => setState('idle'),
      }),
      dropTargetForElements({
        element: el,
        onDrop: () => console.log('dropped on')
      }));

      controller.signal.addEventListener('abort', cleanup, { once: true });
    })();

    return () => {
      controller.abort();
    };
  }, []);

  return <div ref={ref}>Drag me<div>
});
```

## Maximise Deferred

Rather than importing all the modules you need directly, you can pull all your imports and logic out into a seperate file in order to defer more code.

```ts
// attach.js
// regular imports
import {
  draggable,
  dropTargetForElements,
} from '@atlaskit/drag-and-drop/adapter/element';
import { combine } from '@atlaskit/drag-and-drop/util/combine';
import invariant from 'tiny-invariant';

export function attach({ ref, setState }) {
  // all of this code can get deferred
  const el = ref.current;
  invariant(el);
  return combine(
    draggable({
      element: el,
      onDragStart: () => setState('dragging'),
      onDrop: () => setState('idle'),
    }),
    dropTargetForElements({
      element: el,
      onDrop: () => console.log('dropped on'),
    }),
  );
}
```

```tsx
import React, {useEffect} from 'react';

function Card() {
  const ref = useRef<HTMLDivElement | null>(null);
  const [state, setState] = useState();

  useEffect(() => {
    const controller = new AbortController();

    (async () => {
      // just loading in the single module that we need
      const { attach } = import('./attach');
      if (controller.signal.aborted) {
        return;
      }
      const cleanup = attach({setState, ref});

      controller.signal.addEventListener('abort', cleanup, { once: true });
    })();

    return () => {
      controller.abort();
    };
  }, []);

  return <div ref={ref}>Drag me<div>
});
```
