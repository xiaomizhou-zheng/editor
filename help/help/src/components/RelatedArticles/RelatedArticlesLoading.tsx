import React from 'react';
import { injectIntl, WrappedComponentProps } from 'react-intl-next';

import { messages } from '../../messages';

import { DividerLine, LoadingRectangle } from '../../util/styled';

import {
  LoadignRelatedArticleList,
  LoadignRelatedArticleListItem,
} from './styled';

const RelatedArticlesLoading: React.FC<WrappedComponentProps> = ({
  intl: { formatMessage },
}) => {
  return (
    <>
      <LoadignRelatedArticleList
        aria-label={formatMessage(messages.help_loading)}
        role="img"
      >
        <LoadignRelatedArticleListItem>
          <LoadingRectangle
            contentHeight="11px"
            contentWidth="60px"
            marginTop="0"
          />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="8px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
        </LoadignRelatedArticleListItem>

        <LoadignRelatedArticleListItem>
          <LoadingRectangle
            contentHeight="11px"
            contentWidth="60px"
            marginTop="0"
          />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="8px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
        </LoadignRelatedArticleListItem>

        <LoadignRelatedArticleListItem>
          <LoadingRectangle
            contentHeight="11px"
            contentWidth="60px"
            marginTop="0"
          />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="8px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
          <LoadingRectangle contentWidth="100%" marginTop="4px" />
        </LoadignRelatedArticleListItem>
        <DividerLine />
      </LoadignRelatedArticleList>
    </>
  );
};

export default injectIntl(RelatedArticlesLoading);
