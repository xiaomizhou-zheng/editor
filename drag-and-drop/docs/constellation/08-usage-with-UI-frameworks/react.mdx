---
order: 1
title: Usage with UI Frameworks
description: How to use @atlaskit/drag-and-drop with different UI libraries and frameworks
---

### Attaching behaviour

We advise that you attach drag and drop behaviour to elements using [the `useEffect` hook](https://reactjs.org/docs/hooks-effect.html).

```tsx
// card.tsx
import {draggable, dropTargetForElements} from '@atlaskit/drag-and-drop/adapter/element';

export default function Card({ item }: { item: Item }) {
  const ref = useRef<HTMLDivElement | null>(null);
  const itemId = item.itemId;
  const [state, setState] = useState<DraggableState>('idle');

  useEffect(() => {
    const cleanup = combine(
      draggable({
        element: ref.current,
        getInitialData: () => ({ type: 'card', itemId: itemId }),
      }),
      dropTargetForElements({
        element: ref.current,
        canDrop: args => args.source.data.type === 'card',
      }),
    );
    return cleanup;
  }, [itemId]);

  return (
    <div ref={ref}>
      item id: {item.itemId}</span>
    </div>
  );
};
```

It is fine if your `draggable` or `dropTargetForElements` effect is cleaned up and re-created, even during a drag. For both of `draggable` and _drop targets_, the `element` is used as the _key_ for the entity. For more information, see [entity equality](TODO);

### Monitors

Using effects is also a nice way to manage the lifecyle of monitors:

```tsx
export function App() {

  useEffect(() => {
    // this is nice as the monitor will be cleaned up when <App /> is unmounted
    const cleanup = monitorForFiles({
      onDragStart: () => console.log('A file is dragging!');
    });

    return cleanup;
  }, []);

  return <Example>;
};
```

### Deferred loading

`react` lazy loading is based on components ([more details](https://reactjs.org/docs/code-splitting.html)).

You can use this component pattern with `@atlaskit/drag-and-drop`:

```tsx
import {Suspense, lazy} from 'react';

// importing our `card.tsx` file from above
const LazyCard = lazy(() => ('./card.tsx'));

function App() {
  return <Suspense fallback="loading..."><LazyCard></Suspense>
}
```

As `@atlaskit/drag-and-drop` is not tied to `react`, you can also attach drag and drop behaviour at some point after a component has already been rendered:

```tsx
function Card({ item }: { item: Item }) {
  const ref = useRef<HTMLDivElement | null>(null);
  const itemId = item.itemId;

  useEffect(() => {
    const controller = new AbortController();

    // Look! We are attaching behaviour after the component has rendered!
    // Note: Atlassian's, please use http://go/react-async rather than this promise based approach
    (async () => {
      const modules = await Promise.all([
        await import('@atlaskit/drag-and-drop/adapter/element'),
        await import('@atlaskit/drag-and-drop/util/combine'),
      ]);

      if (controller.signal.aborted) {
        return;
      }

      const [{ draggable, dropTargetForElements }, { combine }] = modules;

      if (!ref.current) {
        return;
      }

      const cleanup = combine(
        draggable({
          element: ref.current,
          getInitialData: () => ({ type: 'card', itemId: itemId }),
        }),
        dropTargetForElements({
          element: ref.current,
          canDrop: args => args.source.data.type === 'card',
          getData: () => ({ type: 'card', itemId: itemId }),
        }),
      );

      controller.signal.addEventListener('abort', cleanup, { once: true });
    })();

    return () => {
      controller.abort();
    };
  }, [itemId]);

  return (
    <div ref={ref}>
      item id: {item.itemId}</span>
    </div>
  );
};
```

For more details, see our [deferred loading guide](/components/drag-and-drop/deferred-loading).
