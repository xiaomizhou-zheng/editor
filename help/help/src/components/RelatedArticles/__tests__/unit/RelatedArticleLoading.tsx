import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl-next';

import RelatedArticlesLoading from '../../RelatedArticlesLoading';

describe('RelatedArticlesLoading', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <IntlProvider locale="en">
        <RelatedArticlesLoading />
      </IntlProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
