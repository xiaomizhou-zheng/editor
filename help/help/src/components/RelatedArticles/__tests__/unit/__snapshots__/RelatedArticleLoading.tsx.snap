// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedArticlesLoading Should match snapshot 1`] = `
@keyframes animation-0 {
  0% {
    background-position: -300px 0;
  }

  100% {
    background-position: 1000px 0;
  }
}

@keyframes animation-0 {
  0% {
    background-position: -300px 0;
  }

  100% {
    background-position: 1000px 0;
  }
}

@keyframes animation-0 {
  0% {
    background-position: -300px 0;
  }

  100% {
    background-position: 1000px 0;
  }
}

.emotion-19 {
  width: 100%;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.emotion-5 {
  display: block;
  width: 100%;
  padding: 8px;
  margin-bottom: 16px;
  box-sizing: border-box;
}

.emotion-0 {
  display: inline-block;
  vertical-align: middle;
  position: relative;
  height: 11px;
  margin-top: 0;
  width: 60px;
  border-radius: 2px;
  -webkit-animation-duration: 1.2s;
  animation-duration: 1.2s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-name: animation-0;
  animation-name: animation-0;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  background-color: var(--ds-background-neutral,#EBECF0);
  background-image: linear-gradient( to right, var(--ds-background-neutral-subtle,#EBECF0) 10%, var(--ds-background-neutral,#DFE1E6) 20%, var(--ds-background-neutral-subtle,#EBECF0) 30% );
  background-repeat: no-repeat;
}

.emotion-1 {
  display: inline-block;
  vertical-align: middle;
  position: relative;
  height: 1rem;
  margin-top: 4px;
  width: 100%;
  border-radius: 2px;
  -webkit-animation-duration: 1.2s;
  animation-duration: 1.2s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-name: animation-0;
  animation-name: animation-0;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  background-color: var(--ds-background-neutral,#EBECF0);
  background-image: linear-gradient( to right, var(--ds-background-neutral-subtle,#EBECF0) 10%, var(--ds-background-neutral,#DFE1E6) 20%, var(--ds-background-neutral-subtle,#EBECF0) 30% );
  background-repeat: no-repeat;
}

.emotion-2 {
  display: inline-block;
  vertical-align: middle;
  position: relative;
  height: 1rem;
  margin-top: 8px;
  width: 100%;
  border-radius: 2px;
  -webkit-animation-duration: 1.2s;
  animation-duration: 1.2s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-name: animation-0;
  animation-name: animation-0;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  background-color: var(--ds-background-neutral,#EBECF0);
  background-image: linear-gradient( to right, var(--ds-background-neutral-subtle,#EBECF0) 10%, var(--ds-background-neutral,#DFE1E6) 20%, var(--ds-background-neutral-subtle,#EBECF0) 30% );
  background-repeat: no-repeat;
}

.emotion-18 {
  background-color: var(--ds-border,rgba(9,30,66,0.08));
  height: 2px;
  width: 100%;
  padding: 0 16px;
  margin-top: 16px;
  box-sizing: border-box;
}

<ul
  aria-label="Loading"
  class="emotion-19"
  role="img"
>
  <li
    class="emotion-5"
  >
    <div
      class="emotion-0"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-1"
    />
  </li>
  <li
    class="emotion-5"
  >
    <div
      class="emotion-0"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-1"
    />
  </li>
  <li
    class="emotion-5"
  >
    <div
      class="emotion-0"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-1"
    />
    <div
      class="emotion-1"
    />
  </li>
  <div
    class="emotion-18"
  />
</ul>
`;
